### DISCUSSION TITLE
Reusable useFieldArray incomplete field type

### DISCUSSION THREAD

**Discussion by meckanism** [react-hook-form/react-hook-form#9669] [Q&A] [answered]
Howdy,

Apologies in advance if this ends up being a silly question, my TS foo is not on point yet.
I'm trying to make a reusable `useFieldArray` component but have been struggling to understand why the 'field' type is incomplete.

```js
type FormValues = {
  servers: { ip: string; port: string; desc: string }[];
};

 const { control } = useForm<FormValues>({
    defaultValues: {
      servers: [{ ip: 'a', port: 'b', desc: 'c' }],
    },
  });

  return <SomeComponent<FormValues> control={control} name="servers" />;
```

As you can see in this [example](https://stackblitz.com/edit/react-ts-trcf7g?file=App.tsx) TS says `field.ip` does not exist. 

```js
const SomeComponent = <T extends FieldValues>({
  name,
  control,
}: UseFieldArrayProps<T>) => {
  const { fields } = useFieldArray({
    control,
    name,
  });

  return (
    <div>
      {fields.map((field, index) => {
        console.log(field.ip);
        return <input />;
      })}
    </div>
  );
};
```

What am I doing wrong ?
Any help is appreciated, Cheers.

**Comment by leapful** [comment, marked as answer]
@meckanism 

I think you're mis-understanding generic interface of typescript. 

Applying `<T extends FieldValues>` on your `SomeComponent` only infers type when the component is being used, not inside the component itself, meaning it helps to pass and check type of props into `<SomeComponent<FormValues> control={control} name="servers" />` .

Inside `SomeComponent` implementation, the type refers to `T extends FieldValues` which does not know about the context `field.ip` and it's an expected result.

I would suggest to use a callback to infer type when using the component instead of access it directly inside the component.

Working demo: https://stackblitz.com/edit/react-ts-fzrpvs?file=App.tsx,node_modules%2Freact-hook-form%2Fdist%2FuseFieldArray.d.ts,node_modules%2Freact-hook-form%2Fdist%2Ftypes%2FfieldArray.d.ts

**Reply by meckanism** [reply]
@leapful

I surely am misunderstanding!
Thanks for the great explanation.
